media/client/ipc/source/ControlIpc.cpp

```diff
@@ -0,0 +1,228 @@
+        return false;
+    }
+
+    firebolt::rialto::GetSharedMemoryRequest request;
+
+    firebolt::rialto::GetSharedMemoryResponse response;
+    auto ipcController = m_ipc.createRpcController();
+    auto blockingClosure = m_ipc.createBlockingClosure();
+    controlStub->getSharedMemory(ipcController.get(), &request, &response, blockingClosure.get());
+
+    // wait for the call to complete
+    blockingClosure->wait();
+
+    // check the result
+    if (ipcController->Failed())
+    {
+        RIALTO_CLIENT_LOG_ERROR("failed to get the shared memory due to '%s'", ipcController->ErrorText().c_str());
+        return false;
+    }
+
+    fd = response.fd();
+    size = response.size();
+
+    return true;
+}
+
+bool ControlIpc::registerClient()
+{
+    std::shared_ptr<::firebolt::rialto::ControlModule_Stub> controlStub = m_controlStub;
+
+    if (!reattachChannelIfRequired())
+    {
+        RIALTO_CLIENT_LOG_ERROR("Reattachment of the ipc channel failed, ipc disconnected");
+        return false;
+    }
+
+    firebolt::rialto::RegisterClientRequest request;
+    firebolt::rialto::RegisterClientResponse response;
+    auto ipcController = m_ipc.createRpcController();
+    auto blockingClosure = m_ipc.createBlockingClosure();
+    controlStub->registerClient(ipcController.get(), &request, &response, blockingClosure.get());
+
+    // wait for the call to complete
+    blockingClosure->wait();
+
+    // check the result
+    if (ipcController->Failed())
+    {
+        RIALTO_CLIENT_LOG_ERROR("failed to register client due to '%s'", ipcController->ErrorText().c_str());
+        return false;
+    }
+
+    m_controlHandle = response.control_handle();
+
+    return true;
+}
+
+bool ControlIpc::createRpcStubs()
+{
+    m_controlStub = std::make_unique<::firebolt::rialto::ControlModule_Stub>(m_ipcChannel.get());
+    if (!m_controlStub)
+    {
+        return false;
+    }
+    return true;
+}
+
+bool ControlIpc::subscribeToEvents()
+{
+    if (!m_ipcChannel)
+    {
+        return false;
+    }
+
+    int eventTag = m_ipcChannel->subscribe<firebolt::rialto::ApplicationStateChangeEvent>(
+        [this](const std::shared_ptr<firebolt::rialto::ApplicationStateChangeEvent> &event)
+        { m_eventThread->add(&ControlIpc::onApplicationStateUpdated, this, event); });
+    if (eventTag < 0)
+        return false;
+    m_eventTags.push_back(eventTag);
+
+    eventTag = m_ipcChannel->subscribe<firebolt::rialto::PingEvent>(
+        [this](const std::shared_ptr<firebolt::rialto::PingEvent> &event)
+        { m_eventThread->add(&ControlIpc::onPing, this, event); });
+    if (eventTag < 0)
+        return false;
+    m_eventTags.push_back(eventTag);
+
+    return true;
+}
+
+void ControlIpc::onApplicationStateUpdated(const std::shared_ptr<firebolt::rialto::ApplicationStateChangeEvent> &event)
+{
+    if (m_controlHandle != event->control_handle())
+    {
+        RIALTO_CLIENT_LOG_WARN("ApplicationStateChangeEvent received with wrong handle");
```

Comment:
m_controlHandle is set in ControlIpc::registerClient(), value is taken from RegisterClientResponse. First ApplicationStateChangeEvent may come before this message. I will remove this IF, because that print is not needed.